src/main/java/seedu/address/model/leave/Range.java

```diff
@@ -0,0 +1,102 @@
+package seedu.address.model.leave;
+
+import static java.util.Objects.requireNonNull;
+
+import seedu.address.model.leave.exceptions.EndBeforeStartException;
+
+/**
+ * Represents a range of dates
+ * Guarantees: End date will not be before start date, if both are present
+ */
+public class Range {
+    public static final String MESSAGE_INVALID_END_DATE =
+            "The end date is earlier than the start date!";
+    private final Date startDate;
+    private final Date endDate;
+
+    /**
+     * Constructs a Range object. Since construction is only possible via the static methods,
+     * it is guaranteed that the end date will not be before the start date, if both are present
+     * @param startDate Start date of range/null
+     * @param endDate End date of range/null
+     */
+    private Range(Date startDate, Date endDate) {
+        this.startDate = startDate;
+        this.endDate = endDate;
+    }
```

Comment:
Could our factory method instead handle like a single value (end date passed in and start date defaults as today)? Or should we continue allowing endDate and startDate to be null

Reply:
My concern is that it means the user will now have to check both start date and end date, then figure out which factory method to call, rather than just using the same factory method.

Reply:
perhaps other than startDate and endDate, we can have hasEnd and hasStart here. Such that the predicate can use these boolean values instead of having to handle null in the test method. But logic wise it works great. minor nitpick

Reply:
would it work better if my getStartDate and getEndDate returns optionals instead?